### PR TITLE
REALMC-8240 Add segment tracking to realm-cli

### PR DESCRIPTION
**Pull request by jo-tse-mongo**

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -5,20 +5,23 @@ go 1.14
 require (
 	github.com/AlecAivazis/survey/v2 v2.2.2
 	github.com/Netflix/go-expect v0.0.0-20180615182759-c93bf25de8e8
+	github.com/bmizerany/assert v0.0.0-20160611221934-b7ed37b82869 // indirect
```
Comment (makesitgo) [inline go.mod:8]: i think you should run `go mod tidy` in your project and see what updates here... just curious if `segmentio` needs to bring all these in or if something was included accidentally

File: internal/cli/command_factory.go [hunk 2]
```diff
@@ -151,7 +155,7 @@ func (factory *CommandFactory) SetGlobalFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&factory.profile.atlasBaseURL, flagAtlasBaseURL, "", flagAtlasBaseURLUsage)
 	fs.StringVar(&factory.profile.realmBaseURL, flagRealmBaseURL, "", flagRealmBaseURLUsage)
 	fs.VarP(&factory.profile.telemetryMode, telemetry.FlagMode, telemetry.FlagModeShort, telemetry.FlagModeUsage)
-
+	fs.StringVarP(&factory.profile.telemetryWriteKey, telemetry.FlagWriteKey, telemetry.FlagWriteKeyShort, "", telemetry.FlagWriteKeyUsage)
```
Comment (makesitgo) [inline internal/cli/command_factory.go:158]: this should not be a global flag, we can remove this

File: internal/cli/command_factory.go [hunk 3]
```diff
@@ -76,11 +76,15 @@ func (factory *CommandFactory) Build(command CommandDefinition) *cobra.Command {
 				os.Exit(1)
 			}
 
-			factory.telemetryService = telemetry.NewService(
+			telemetryService := telemetry.NewService(
 				factory.profile.telemetryMode,
 				factory.profile.User().PublicAPIKey,
+				factory.profile.telemetryWriteKey,
```
Comment (makesitgo) [inline internal/cli/command_factory.go:82]: can remove this, we are going to store the write key as a `var` in the `telemetry` package

File: internal/cli/command_factory.go [hunk 4]
```diff
@@ -76,11 +76,15 @@ func (factory *CommandFactory) Build(command CommandDefinition) *cobra.Command {
 				os.Exit(1)
 			}
 
-			factory.telemetryService = telemetry.NewService(
+			telemetryService := telemetry.NewService(
```
Comment (makesitgo) [inline internal/cli/command_factory.go:79]: I'm guessing you originally changed this b/c you introduced an error return value to `NewService`, but since thats not needed can we revert this back to how it was?

File: internal/cli/profile.go [hunk 5]
```diff
@@ -133,6 +135,15 @@ func (p *Profile) resolveFlags() error {
 	}
 	p.SetString(keyTelemetryMode, string(p.telemetryMode))
 
+	if p.telemetryWriteKey == "" {
```
Comment (makesitgo) [inline internal/cli/profile.go:138]: all of this write key stuff should not be part of the profile, it should not be configurable by the user (they should only be able to flip the telemetry mode here)

File: internal/telemetry/event.go [hunk 6]
```diff
@@ -37,3 +37,11 @@ type EventDataKey string
 const (
 	EventDataKeyErr EventDataKey = "err"
 )
+
+func (e *event) createPropertyMap() map[string]interface{} {
```
Comment (makesitgo) [inline internal/telemetry/event.go:41]: lets make this a bit more go-like:

```go
func eventDataProperties(data []EventData) map[string]interface{} {
  // func impl here
}
```

File: internal/telemetry/flags.go [hunk 7]
```diff
@@ -12,6 +12,10 @@ const (
 	FlagMode      = "telemetry"
 	FlagModeShort = "m"
 	FlagModeUsage = `enable or disable telemetry (this setting is remembered), available options: ["off", "on"]`
+
+	FlagWriteKey      = "write_key"
```
Comment (makesitgo) [inline internal/telemetry/flags.go:16]: ultimately these should be able to be removed as well

File: internal/telemetry/service.go [hunk 8]
```diff
@@ -26,7 +27,7 @@ func NewService(mode Mode, userID string, command string) *Service {
 	case ModeOff:
 		service.tracker = &noopTracker{}
 	case ModeEmpty, ModeOn:
-		service.tracker = &segmentTracker{}
+		service.tracker = newSegmentTracker(writeKey, logger)
```
Comment (makesitgo) [inline internal/telemetry/service.go:30]: the constructor here should only take in the `logger`

File: internal/telemetry/tracker.go [hunk 9]
```diff
@@ -26,15 +30,24 @@ func (tracker *stdoutTracker) Track(event event) {
 	)
 }
 
-type segmentTracker struct{}
+type segmentTracker struct {
+	client analytics.Client
+	logger *log.Logger
+}
+
+func newSegmentTracker(writeKey string, logger *log.Logger) *segmentTracker {
```
Comment (makesitgo) [inline internal/telemetry/tracker.go:38]: ok, so here is where you're going to want:

```go
var (
  segmentWriteKey = "" // will substitute value at build time with proper key
)
```

and then you can just refer to that when you call `analytics.New`

Comment (makesitgo) [reply internal/telemetry/tracker.go:38]: still trying to think about what we want to do to protect ourselves if this build substitution doesn't happen properly ... 

i think we maybe re-introduce that `error` return value in the telemetry service constructor, but lets discuss this more offline since you have a better idea of the nuances here.  i just want to make sure failed telemetry doesn't prevent the CLI from running at all

**Review by makesitgo — COMMENTED** [review]
just a quick first pass here, lets sync up tomorrow about that error handling

File: go.mod (hunk 1 above)
Comment (jo-tse-mongo) [reply go.mod:8]: I did `go mod tidy`, but I didn't see any changes to commit

File: internal/telemetry/tracker.go (hunk 9 above)
Comment (jo-tse-mongo) [reply internal/telemetry/tracker.go]: I didn't reintroduce the error for the constructor for now and made it return another tracker instead, since it's not really an error that we should fail the CLI for. I'm unsure how the `stdout` tracker would be used (other than maybe dev), but maybe we could return that instead of the `segment` tracker

Comment (makesitgo) [reply internal/telemetry/tracker.go]: the `stdout` tracker may go away all together... its noisy and and not all that helpful for anyone involved haha (it was cool during early developemnt).

i like the idea of just transitioning over to use the `noopTracker` if we find that `segmentWriteKey` is empty, let's roll with that

File: internal/cli/command_factory.go [hunk 10]
```diff
@@ -151,7 +152,6 @@ func (factory *CommandFactory) SetGlobalFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&factory.profile.atlasBaseURL, flagAtlasBaseURL, "", flagAtlasBaseURLUsage)
 	fs.StringVar(&factory.profile.realmBaseURL, flagRealmBaseURL, "", flagRealmBaseURLUsage)
 	fs.VarP(&factory.profile.telemetryMode, telemetry.FlagMode, telemetry.FlagModeShort, telemetry.FlagModeUsage)
-
```
Comment (makesitgo) [inline internal/cli/command_factory.go:154]: unnecessary diff here

File: internal/cli/profile.go [hunk 11]
```diff
@@ -34,9 +34,10 @@ const (
 
 // Profile is the CLI profile
 type Profile struct {
-	Name          string
-	atlasBaseURL  string
-	realmBaseURL  string
+	Name         string
+	atlasBaseURL string
+	realmBaseURL string
+
```
Comment (makesitgo) [inline internal/cli/profile.go:40]: unnecessary diff here as well

File: internal/telemetry/service_test.go [hunk 12]
```diff
@@ -9,8 +9,7 @@ import (
 
 func TestNewService(t *testing.T) {
 	t.Run("Should create the expected Service", func(t *testing.T) {
-		service := NewService(ModeOn, "userID", "command")
-
```
Comment (makesitgo) [inline internal/telemetry/service_test.go:13]: keep the newline here, in general I'd suggest trying to keep the amount of diffs only to what's necessary.  the newline stuff (adding/removing) just adds noise to the PRs

File: internal/telemetry/tracker.go [hunk 13]
```diff
@@ -2,7 +2,15 @@ package telemetry
...
 	"fmt"
+	"log"
 	"time"
+
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"gopkg.in/segmentio/analytics-go.v3"
+)
+
+var (
+	segmentWriteKey = ""
```
Comment (makesitgo) [inline internal/telemetry/tracker.go:13]:
```suggestion
	segmentWriteKey = "" // value will be injected at build-time
```

File: internal/telemetry/tracker.go [hunk 14]
```diff
@@ -26,15 +34,28 @@ func (tracker *stdoutTracker) Track(event event) {
...
 
-type segmentTracker struct{}
+type segmentTracker struct {
+	client analytics.Client
+	logger *log.Logger
+}
+
+func newSegmentTracker(logger *log.Logger) Tracker {
+	if len(segmentWriteKey) == 0 {
+		log.Print("unable to make a Segment tracker with an empty write key; replacing with a noop")
```
Comment (makesitgo) [inline internal/telemetry/tracker.go:44]:
```suggestion
		log.Print("unable to connect to Segment due to missing key, CLI telemetry will be disabled")
```

File: internal/telemetry/tracker.go [hunk 15]
```diff
@@ -26,15 +34,28 @@ func (tracker *stdoutTracker) Track(event event) {
...
 func (tracker *segmentTracker) Track(event event) {
-	fmt.Printf(
-		"%s UTC TELEM %s: %s%v\n",
-		event.time.In(time.UTC).Format("15:04:05"),
-		event.command,
-		event.eventType,
-		event.data,
-	)
+	if err := tracker.client.Enqueue(analytics.Track{
+		MessageId:  primitive.NewObjectID().Hex(),
```
Comment (makesitgo) [inline internal/telemetry/tracker.go:53]: use `event.id` here

File: internal/telemetry/tracker.go [hunk 16]
```diff
@@ -26,15 +34,28 @@ func (tracker *stdoutTracker) Track(event event) {
...
+		Timestamp:  time.Now(),
```
Comment (makesitgo) [inline internal/telemetry/tracker.go:54]: use `event.time` here

File: internal/telemetry/tracker.go [hunk 17]
```diff
@@ -26,15 +34,28 @@ func (tracker *stdoutTracker) Track(event event) {
...
+		Event:      string(event.eventType),
+		UserId:     event.userID,
+		Properties: eventDataProperties(event.data),
```
Comment (makesitgo) [inline internal/telemetry/tracker.go:57]: I might have sent you down a bad path here, but we probably don't even need `eventDataProperties` as a function (and could/should handle that logic just here in this function).

the thing that's missing right now are two additional `EventData`s we want to care about:

```go
[]EventData{
  {"cmd", event.command}, // "cmd" => eventDataKeyCommand
  {"xid", event.executionID}, // "xid" => eventDataKeyExecutionID
}
```

since this is all just segment-specific stuff, i think using `make(map[string]interface{}, len(event.data)+2)` inside this function before the `client.Enqueue` call works fine here

File: internal/telemetry/tracker_test.go [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+func (client *testClient) Close() error {
+	return nil
+}
+
+var testTime = time.Date(2021, 1, 2, 3, 4, 5, 6, time.UTC)
+
+func TestNoopTracker(t *testing.T) {
+	t.Run("should create a noop tracker", func(t *testing.T) {
+		tracker := noopTracker{}
+		assert.NotNil(t, tracker)
```
Comment (makesitgo) [inline internal/telemetry/tracker_test.go:31]: this assert is not doing much here, we can probably get rid of this outer test-scope here

File: internal/telemetry/tracker_test.go [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+			testEvent := createEvent(EventTypeCommandStart, nil, "someCommand")
+			testTrackStdoutOutput(t, &tracker, testEvent, "")
+		})
+	})
+}
+
+func TestStdoutTracker(t *testing.T) {
+	t.Run("should create an stdout tracker", func(t *testing.T) {
+		tracker := stdoutTracker{}
+		assert.NotNil(t, tracker)
```
Comment (makesitgo) [inline internal/telemetry/tracker_test.go:42]: same here for getting rid of the assert

File: internal/telemetry/tracker_test.go [hunk 20]
```diff
@@ -0,0 +1,101 @@
...
+		})
+	})
+}
+
+func TestSegmentTracker(t *testing.T) {
+	client := &testClient{}
+	t.Run("should create the segment tracker", func(t *testing.T) {
+		tracker := segmentTracker{}
+		tracker.client = client
+		assert.NotNil(t, tracker)
```
Comment (makesitgo) [inline internal/telemetry/tracker_test.go:59]: same thing here

File: internal/telemetry/tracker_test.go [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+		t.Run("and should print the tracking information to the logger", func(t *testing.T) {
+			tracker.Track(createEvent(EventTypeCommandError, []EventData{{Key: EventDataKeyErr, Value: "Something"}}, "someCommand"))
+			testClientResults := tracker.client.(*testClient)
+			assert.Equal(t, 1, len(testClientResults.calls))
+
+			actualTrack := testClientResults.calls[0].(analytics.Track)
+			assert.NotNil(t, actualTrack)
+			assert.Equal(t, string(EventTypeCommandError), actualTrack.Event)
```
Comment (makesitgo) [inline internal/telemetry/tracker_test.go:67]: would we be able to have all of this in a single `assert.Equal` by building an `analytics.Track` ourselves?

something like: `assert.Equal(t, []interface{}{analytics.Track{ ... }}, testClientResults.calls)`

**Review by makesitgo — COMMENTED** [review]
lookin real close here, i want to dive a little deeper into that `testTrackStdoutOutput` func (looks clever!) before giving the full 👍 ... here's my last bit of cleanup comments

File: internal/telemetry/tracker.go [hunk 22]
```diff
@@ -26,15 +33,32 @@ func (tracker *stdoutTracker) Track(event event) {
...
-// TODO(REALMC-7243): use Segment sdk to send events through client
 func (tracker *segmentTracker) Track(event event) {
-	fmt.Printf(
-		"%s UTC TELEM %s: %s%v\n",
-		event.time.In(time.UTC).Format("15:04:05"),
-		event.command,
-		event.eventType,
-		event.data,
-	)
+	properties := make(map[string]interface{}, len(event.data))
```
Comment (makesitgo) [inline internal/telemetry/tracker.go:51]: btw we still want to be adding `event.command` and `event.executionID` as properties here (see my previous comment re: `"cmd"` and `"xid"` event data keys

File: internal/telemetry/tracker_test.go [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+		expectedTrack := analytics.Track{
+			MessageId: testID,
+			UserId:    testUserID,
+			Timestamp: testTime,
+			Event:     string(EventTypeCommandError),
+			Properties: map[string]interface{}{
+				string(EventDataKeyErr): "Something",
+			},
+		}
+		actualTrack := testClientResults.calls[0].(analytics.Track)
```
Comment (makesitgo) [inline internal/telemetry/tracker_test.go:72]: out of curiosity, was something failing when you tried to run:

```go
assert.Equal(t, []interface{}{expectedTrack}, testClientResults.calls)
```

If something like that works let's use it, I'd like to keep the asserts as simple as possible.

If I'm missing something there and that doesn't work, let's at least avoid any potential panics within this test.  You should always try to use the two-return-value `, ok` syntax in tests and `assert.True` on the `ok` var (this gives us a nicer test failure message than panics will).  Also, the last assert here ("equals" on the two tracks) should cover the two asserts before that ("non nil" and "length of properties is 1")

File: internal/telemetry/tracker_test.go [hunk 24]
```diff
@@ -0,0 +1,103 @@
...
+		assert.NotNil(t, actualTrack)
+		assert.Equal(t, 1, len(actualTrack.Properties))
+		assert.Equal(t, expectedTrack, actualTrack)
+	})
+}
+
+func testTrackStdoutOutput(t *testing.T, tracker Tracker, event event, expected string) {
```
Comment (makesitgo) [inline internal/telemetry/tracker_test.go:79]: left you a comment about moving this code from a helper to the test case itself, but here are a few other comments to clean up:
* always check the error (you're ignoring errors from `os.Pipe`, `w.Close()`, & `ioutil.ReadAll` here), you have `t *testing.T` here so you can `assert.Nil` easily
* in general, when using helper functions that pass around `t *testing.T` (presumably so you can make `assert` calls), you'll want to make sure to call `t.Helper()` at the top/first thing of the helper function (it ensures proper test output if/when a test fails)
  * fwiw, if/when you move this code to the test func itself, `t.Helper()` will become unnecessary
* b/c you have the defer which replaces `os.Stdout` with the original reference of it, you won't need to do it again on L#89 here

**Review by makesitgo — COMMENTED** [review]
last bit of thoughts here, once you clean it all up this should be good to go. nice work!

File: internal/telemetry/tracker.go (hunk 22 above)
Comment (jo-tse-mongo) [reply internal/telemetry/tracker.go:51]: Is there a reason why `event.Data` is an array? Can it just be `"err"`? It seems strange that `cmd` and `xid` are taken from the event at a different level than `err` is (on the `event` object itself vs within the `data` property)

Comment (makesitgo) [reply internal/telemetry/tracker.go]: if you look at the signature for `(*telemetry.Service).TrackEvent`, you'll see the minimal piece of information needed by the calling code is the `EventType` to track on, and it optionally supports any number of additional `EventData` tuples/structs/objects

for the "start" and "complete" events, we don't really care about any extra details, but on the "error" event we leverage that ability to include an error message

otherwise, the properties that we want to be able to see in every event tracked are provided at the time when we create the `telemetry.Service` (user id, command name).  those are known ahead of time, so we avoid passing them in every time to the `start`/`error`/`complete` stages

Comment (jo-tse-mongo) [reply internal/telemetry/tracker.go]: Should I add `cmd` and `xid` as `EventDataKey`s ?

Comment (makesitgo) [reply internal/telemetry/tracker.go]: we can add them as `EventDataKey`, but no need to export them:

```go
// set of event data keys
const (
	eventDataKeyCommand     EventDataKey = "cmd"
	eventDataKeyExecutionID EventDataKey = "xid"

	EventDataKeyErr EventDataKey = "err"
)
```

File: internal/telemetry/tracker_test.go [hunk 25]
```diff
@@ -0,0 +1,108 @@
...
+			t,
+			&tracker,
+			createEvent(EventTypeCommandComplete, nil, "someCommand"),
+			"03:04:05 UTC TELEM someCommand: COMMAND_COMPLETE[]\n",
+		)
+	})
+}
+
+func TestSegmentTracker(t *testing.T) {
+	client := &testClient{}
```
Comment (makesitgo) [inline internal/telemetry/tracker_test.go:55]: [nit] move this down into the test scope its used in

File: internal/telemetry/tracker_test.go [hunk 26]
```diff
@@ -0,0 +1,108 @@
...
+	t.Run("should create the segment tracker and should print the tracking information to the logger", func(t *testing.T) {
+		tracker := segmentTracker{}
+		tracker.client = client
+		tracker.Track(createEvent(EventTypeCommandError, []EventData{{Key: EventDataKeyErr, Value: "Something"}}, "someCommand"))
+
+		testClientResults := tracker.client.(*testClient)
```
Comment (makesitgo) [inline internal/telemetry/tracker_test.go:61]: you don't need this type cast, you can just use the ref to `client` above

File: internal/telemetry/tracker_test.go [hunk 27]
```diff
@@ -0,0 +1,108 @@
...
+			Event:     string(EventTypeCommandError),
+			Properties: map[string]interface{}{
+				string(EventDataKeyErr):         "Something",
+				string(eventDataKeyCmd):         "someCommand",
+				string(eventDataKeyExecutionID): testExecutionID,
+			},
+		}
+
+		actualTrack := testClientResults.calls[0].(analytics.Track)
+		assert.NotNil(t, actualTrack)
```
Comment (makesitgo) [inline internal/telemetry/tracker_test.go:75]: the stuff with `actualTrack` should go away now, we don't need it

**Review by makesitgo — COMMENTED** [review]
_this_ is the last bit now, i promise 😬

File: internal/telemetry/event.go [hunk 28]
```diff
@@ -30,10 +30,10 @@ const (
 	EventTypeCommandError    EventType = "COMMAND_ERROR"
 )
 
-// EventDataKey is an event data key
-type EventDataKey string
-
 // set of event data keys
 const (
-	EventDataKeyErr EventDataKey = "err"
+	eventDataKeyCommand     string = "cmd"
```
Comment (makesitgo) [inline internal/telemetry/event.go:35]: [nit] shouldn't need the `string` types here (the `const` will pick that up automatically)

File: internal/telemetry/event.go [hunk 29]
```diff
@@ -30,10 +30,10 @@ const (
...
+	eventDataKeyExecutionID string = "xid"
+
+	EventDataKeyErr string = "err"
```
Comment (makesitgo) [inline internal/telemetry/event.go:38]: [driveby fix] rename to `EventDataKeyError`

File: internal/telemetry/service_test.go [hunk 30]
```diff
@@ -47,3 +82,20 @@ type testTracker struct {
 func (tracker *testTracker) Track(event event) {
 	tracker.lastTrackedEvent = event
 }
+func testServiceStdoutput(t *testing.T, createFn func() *Service, expected string) {
```
Comment (makesitgo) [inline internal/telemetry/service_test.go:85]: `Stdoutput`, while an excellent _portmanteau_, isn't exactly meaningful in this context... let's refactor this to be:

```go
func testServiceOutput(t *testing.T, service *Service, expectedOutput string) {
  t.Helper()

  // rest of code here
}
```

File: internal/telemetry/service_test.go [hunk 31]
```diff
@@ -2,14 +2,49 @@ package telemetry
...
-		service := NewService(ModeOn, "userID", "command")
+		service := NewService(ModeStdout, "userID", nil, "command")
+
+		assert.Equal(t, "command", service.command)
+		assert.True(t, service.executionID != "", "service execution id must not be blank")
+		assert.Equal(t, "userID", service.userID)
+		assert.NotNil(t, service.tracker)
+	})
+
+	createSegmentFn := func() *Service {
```
Comment (makesitgo) [inline internal/telemetry/service_test.go:23]: to tack on with the `testServiceOutput` refactor I suggested below, I'll channel my inner-Tim here and resist the one-liner helper func.  instead, we can have:

```go
const (
  testUser = "userID"
  testCommand = "command"
)

var (
  testLogger = log.New(os.Stdout, "LogPrefix ", log.Lmsgprefix)
)

// and instead of passing `createSegmentFn` around here, just instantiate a `*Service` in the tests directly and use those

// and if you *really* want a helper for this still, then go with:
func newService(mode Mode) *Service {
  return NewService(mode, testUser, testLogger, testCommand) // then make sure to be using this for all `NewService` calls
}
```

File: internal/telemetry/service_test.go [hunk 32]
```diff
@@ -2,14 +2,49 @@ package telemetry
...
+	t.Run("Should create a segment tracking service if the segmentWriteKey is there", func(t *testing.T) {
+		swk := segmentWriteKey
+		defer func() { segmentWriteKey = swk }()
+
+		segmentWriteKey = "testing"
+		testServiceStdoutput(t, createSegmentFn, "")
+
+		service := createSegmentFn()
+
+		assert.Equal(t, "command", service.command)
```
Comment (makesitgo) [inline internal/telemetry/service_test.go:35]: these asserts are already verified with `ModeStdout` ... we don't need them here as well (since the logic that handles setting these details is independent of whichever `mode` is passed in)

File: internal/telemetry/tracker_test.go [hunk 33]
```diff
@@ -0,0 +1,104 @@
...
+				EventDataKeyErr:         "Something",
+				eventDataKeyCommand:     "someCommand",
+				eventDataKeyExecutionID: testExecutionID,
+			},
+		}
+		assert.Equal(t, []interface{}{expectedTrack}, client.calls)
+	})
+}
+
+func testTrackStdoutOutput(t *testing.T, tracker Tracker, event event, expected string) {
```
Comment (makesitgo) [inline internal/telemetry/tracker_test.go:76]: lets align this with the other similar helper here:

```go
func testTrackerOutput(t *testing.T, tracker Tracker, expectedOutput string) {
  t.Helper()

  // no need to pass `event` in here, just use a constant, test event and the expectedOutput should still match accordingly
}
```

File: internal/telemetry/tracker_test.go [hunk 34]
```diff
@@ -0,0 +1,104 @@
...
+			Timestamp: testTime,
+			Event:     string(EventTypeCommandError),
+			Properties: map[string]interface{}{
+				EventDataKeyErr:         "Something",
+				eventDataKeyCommand:     "someCommand",
+				eventDataKeyExecutionID: testExecutionID,
+			},
+		}
+		assert.Equal(t, []interface{}{expectedTrack}, client.calls)
+	})
```
Comment (makesitgo) [inline internal/telemetry/tracker_test.go:73]: not sure why i missed this until now, but we want a test here to verify the error case works as expected as well

basically, the same approach you took with `testClient`, but instead of capturing `client.calls` you'll just return an error.

while you add this, lets also rework the names a bit to avoid any ambiguity/confusion.  the `testClient` you currently have should be come `mockSegmentClient`, and this new one can be `mockFailureSegmentClient`

**Review by makesitgo — COMMENTED** [review]
just a bit more test clean up... and after the suggestions are up here we'll be good to go

**Review by makesitgo — APPROVED** [review]
🆒  beans
